packages/core/expressions/src/components/ExpressionsEditor.vue

```diff
@@ -0,0 +1,211 @@
+import type * as Monaco from 'monaco-editor'
+import * as monaco from 'monaco-editor'
+import { computed, onBeforeUnmount, onMounted, ref, watch } from 'vue'
+import { createSchema, type NamedSchemaDefinition } from '../schema'
+import { registerLanguage, registerTheme, theme } from '../monaco'
+
+let editor: Monaco.editor.IStandaloneCodeEditor | undefined
+let editorModel: Monaco.editor.ITextModel
+
+const parse = (expression:string, schema: Schema) => Parser.parse(expression, schema)
+
+const { debounce } = useDebounce()
+
+const props = withDefaults(defineProps<{
+  schema: NamedSchemaDefinition,
+  parseDebounce?: number,
+  inactiveUntilFocused?: boolean,
+}>(), {
+  parseDebounce: 500,
+})
+
+const expression = defineModel<string>({ required: true })
+
+const emit = defineEmits<{
+  'parse-result-update': [result: ParseResult]
+}>()
+
+const root = ref(null)
+const isParsingActive = ref(false)
+const parseResult = ref<ParseResult | undefined>()
+
+const editorClass = computed(() => [
+  'expression-editor',
+  { invalid: isParsingActive.value && parseResult.value?.status !== 'ok' },
+])
+
+const registerSchema = (schema: NamedSchemaDefinition) => {
+  const { languageId } = registerLanguage(schema)
+  monaco.editor.setModelLanguage(editorModel, languageId)
+}
+
+onMounted(() => {
+  registerTheme()
+
+  editor = monaco.editor.create(root.value!, {
+    automaticLayout: true,
+    fixedOverflowWidgets: true,
+    fontSize: 14,
+    lineNumbersMinChars: 3,
+    lineDecorationsWidth: 2,
+    minimap: {
+      enabled: false,
+    },
+    renderValidationDecorations: 'editable',
+    overviewRulerLanes: 0,
+    renderLineHighlightOnlyWhenFocus: true,
+    scrollBeyondLastLine: false,
+    theme,
+    value: expression.value,
+  })
+
+  editor.onDidChangeModelContent(() => {
+    const value = editor!.getValue()
+    expression.value = value
+  })
+
+  editorModel = editor.getModel()!
+
+  if (props.inactiveUntilFocused) {
+    editor.onDidFocusEditorWidget(() => {
+      if (!isParsingActive.value) {
+        registerSchema(props.schema)
+        isParsingActive.value = true
+        parseResult.value = parse(expression.value, createSchema(props.schema.definition))
+      }
+    })
+  } else {
+    registerSchema(props.schema)
+    isParsingActive.value = true
+    parseResult.value = parse(expression.value, createSchema(props.schema.definition))
+  }
+})
+
+onBeforeUnmount(() => {
+  editor?.dispose()
+})
+
+const schema = computed(() => createSchema(props.schema.definition))
+
+watch(() => props.schema, (newSchema) => {
+  const { languageId } = registerLanguage(newSchema)
+  monaco.editor.setModelLanguage(editorModel, languageId)
+}, { deep: true })
+
+watch(expression, (newExpression) => {
+  if (!isParsingActive.value) {
+    isParsingActive.value = true
+  }
+
+  if (editor !== undefined && editor.getValue() !== newExpression) {
+    editor.setValue(newExpression)
+  }
+})
+
+watch([expression, schema], (() => {
+  const cb = () => {
```

Comment:
nit: could a more descriptive variable name be used or maybe add a comment explaining what this watch is doing?